File: tree/1991_TreeTraversal.py
class TreeNode:
    def __init__(self,val,left=None,right=None):
        self.val=val
        self.left=left
        self.right=right

    def push(self,head,parent,val1,val2):
        if head:
            if head.val==parent:
                if val1!='.':
                    head.left=TreeNode(val1)
                if val2!='.':
                    head.right=TreeNode(val2)
            else:
                head.left=self.push(head.left,parent,val1,val2)
                head.right=self.push(head.right,parent,val1,val2)
        
        return head

    def preorder(self,head):
        if head:
            print(head.val,end="")
            self.preorder(head.left)
            self.preorder(head.right)

    def inorder(self,head):
        if head:
            self.inorder(head.left)
            print(head.val,end="")
            self.inorder(head.right)

    def postorder(self,head):
        if head:
            self.postorder(head.left)
            self.postorder(head.right)
            print(head.val,end="")
                
        
head=TreeNode('A')
for _ in range(int(input())):
    p,c1,c2=input().split()
    head=head.push(head,p,c1,c2)

head.preorder(head); print()
head.inorder(head); print()
head.postorder(head); print()

File: tree/1991_TreeTraversal_.cpp
#include<iostream>
#include<vector>
#define MAX 26

using namespace std;

vector<int>tree[MAX];

void preorder(int n) {
	if (n >= 0) {
		cout << char(n + 'A');
		preorder(tree[n][0]);
		preorder(tree[n][1]);
	}
}

void inorder(int n) {
	if (n >= 0) {
		inorder(tree[n][0]);
		cout << char(n + 'A');
		inorder(tree[n][1]);
	}
}

void postorder(int n) {
	if (n >= 0) {
		postorder(tree[n][0]);
		postorder(tree[n][1]);
		cout << char(n + 'A');
	}
}

int main() {
	int N;
	cin >> N;
	
	for (int i = 0; i < N; i++) {
		char p, c1, c2;
		cin >> p >> c1 >> c2;
		
		tree[p - 'A'].push_back(c1-'A'); // .이면 음수를 추가함
		tree[p - 'A'].push_back(c2 - 'A');
	}
	
	preorder(0); cout << "\n";
	inorder(0); cout << "\n";
	postorder(0); cout << "\n";

	return 0;
}
